keps/sig-scheduling/5055-dra-device-taints-and-tolerations/README.md

```diff
@@ -258,30 +306,85 @@ Device and node taints are applied independently. A node taint applies to all
+the eviction controller to calculate which devices are matched and which pods
+would need to get evicted. The input for that code is the single
+DeviceTaintRule which gets described and the complete set of ResourceSlices,
+ResourceClaims and Pods.
+
+Calculating the outcome on the client side was chosen because the alternative,
+always updating the status of DeviceTaintRules on the server side, would cause
+permanent churn regardless whether someone wants the information or not.
+
+The exact output will be decided during the implementation phase.
+It can change at any time because the `kubectl describe` output does not fall
+under any Kubernetes API stability guarantees.
+
 ### API
 
-The ResourceSlice content gets extended:
+The ResourceSlice content gets extended. A natural place for a taint that
+affects one specific device would be inside the `Device` struct. The problem
+with that approach is that with 128 allowed devices per ResourceSlice the
+maximum size of each taint would have to be very small to keep the worst-case
+size of a ResourceSlice within the limits imposed by etcd and the API request
+machinery.
+
+Therefore a different approach is used where each ResourceSlice either provides
+information about devices or about taints, but never both. Using the
+ResourceSlice for taints instead of a different type has the advantage that the
+existing mechanisms and code for publishing and consuming the information can
+be reused. The same approach is likely to be used for other additional
+ResourceSlice pool information, like mixin definitions. The downside is that at
+least two ResourceSlices are necessary once taints get published by a DRA
+driver.
 
 ```Go
-type Device struct {
+type ResourceSliceSpec struct {
     ...
 
-    // If specified, these are the driver-defined taints.
+    // Devices lists some or all of the devices in this pool.
     //
-    // The maximum number of taints is 4.
+    // Must not have more than 128 entries. Either Devices or Taints may be set, but not both.
+    //
+    // +optional
+    // +listType=atomic
+    // +oneOf=ResourceSliceContent
+    Devices []Device
+
+    // If specified, these are driver-defined taints.
+    //
+    // The maximum number of taints is 32. Either Devices or Taints may be set, but not both.
     //
     // This is an alpha field and requires enabling the DRADeviceTaints
     // feature gate.
     //
     // +optional
     // +listType=atomic
     // +featureGate=DRADeviceTaints
-    Taints []DeviceTaint
+    // +oneOf=ResourceSliceContent
+    Taints []SliceDeviceTaint
+}
+
+// DeviceTaintsMaxLength is the maximum number of taints per ResourceSlice.
+const DeviceTaintsMaxLength = 32
+
+// SliceDeviceTaint defines one taint within a ResourceSlice.
+type SliceDeviceTaint struct {
+   // Device is the name of the device in the pool that the ResourceSlice belongs to
+   // which is affected by the taint. Multiple taints may affect the same device.
+   Device string
+
+   DeviceTaint
```

Comment:
My initial reasoning for embedding here was that a "SliceDeviceTaint is a DeviceTaint". But "has a" also makes sense. Both works for the code.

I'm now leaning towards calling this field `Taint DeviceTaint`, because that makes it symmetric with
```
type DeviceTaintRuleSpec struct {
    DeviceSelector *DeviceTaintSelector
    Taint DeviceTaint
}
```

Reply:
I think I also slightly prefer the consistency of the non-embedded field. I'm happy to take on the burden of changing that if that would be helpful.